semantic-core/src/Data/Term.hs

```diff
@@ -0,0 +1,93 @@
+{-# LANGUAGE DeriveTraversable, FlexibleInstances, LambdaCase, MultiParamTypeClasses, QuantifiedConstraints, RankNTypes, ScopedTypeVariables, StandaloneDeriving, TypeOperators, UndecidableInstances #-}
+module Data.Term
+( Term(..)
+, Syntax(..)
+, iter
+, cata
+, interpret
+) where
+
+import Control.Effect.Carrier
+import Control.Monad (ap)
+import Control.Monad.Module
+import Data.Coerce (coerce)
+import Data.Functor.Const
+import Data.Scope
+
+data Term sig a
+  = Var a
+  | Term (sig (Term sig) a)
+
+deriving instance ( Eq a
+                  , RightModule sig
+                  , forall g x . (Eq  x, Monad g, forall y . Eq  y => Eq  (g y)) => Eq  (sig g x)
+                  )
+               => Eq  (Term sig a)
+deriving instance ( Ord a
+                  , RightModule sig
+                  , forall g x . (Eq  x, Monad g, forall y . Eq  y => Eq  (g y)) => Eq  (sig g x)
+                  , forall g x . (Ord x, Monad g, forall y . Eq  y => Eq  (g y)
+                                                , forall y . Ord y => Ord (g y)) => Ord (sig g x)
+                  )
+               => Ord (Term sig a)
+deriving instance (Show a, forall g x . (Show x, forall y . Show y => Show (g y)) => Show (sig g x)) => Show (Term sig a)
+
+deriving instance ( forall g . Foldable    g => Foldable    (sig g)) => Foldable    (Term sig)
+deriving instance ( forall g . Functor     g => Functor     (sig g)) => Functor     (Term sig)
+deriving instance ( forall g . Foldable    g => Foldable    (sig g)
+                  , forall g . Functor     g => Functor     (sig g)
+                  , forall g . Traversable g => Traversable (sig g)) => Traversable (Term sig)
```

Comment:
We don’t get an infinite loop from using more specialized constraints for these instances:

```haskell
deriving instance Foldable    (sig (Term sig)) => Foldable    (Term sig)
deriving instance Functor     (sig (Term sig)) => Functor     (Term sig)
deriving instance Traversable (sig (Term sig)) => Traversable (Term sig)
```

However, these are exactly the sort of undecidable instances that quantified constraints were designed to avoid, and given that we’ve already courted undecidability above, I felt it best to play it safe.